File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.py
from collections import defaultdict
from math import ceil


class Fee:
    def __init__(self, base_time, base_fee, unit_time, unit_fee):
        self.base_time = base_time
        self.base_fee = base_fee
        self.unit_time = unit_time
        self.unit_fee = unit_fee

    def calculate_cost(self, time):
        fee = self.base_fee

        exceed_time = max(0, time - self.base_time)
        fee += ceil(exceed_time / self.unit_time) * self.unit_fee

        return fee


class Car:
    def __init__(self):
        self.last_in_time = 0
        self.total_time = 0

    def car_in(self, in_time):
        self.last_in_time = in_time

    def car_out(self, out_time):
        if self.last_in_time == -1:
            return

        self.total_time += (out_time - self.last_in_time)
        self.last_in_time = -1

    def get_cost(self, fee):
        return fee.calculate_cost(self.total_time)


def parse_time(s):
    hour, minute = map(int, s.split(":"))
    return hour * 60 + minute


def solution(fees, records):
    fee = Fee(fees[0], fees[1], fees[2], fees[3])
    cars = defaultdict(lambda: Car())

    for record in records:

        times, car_number, status = record.split()
        car_time = parse_time(times)

        if status == "IN":
            cars[car_number].car_in(car_time)
        else:
            cars[car_number].car_out(car_time)

    end_time = parse_time("23:59")
    for car in cars.values():
        car.car_out(end_time)

    ans = []

    for car_number in sorted(cars):
        ans.append(cars[car_number].get_cost(fee))

    return ans

File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.java
import java.util.*;

class Solution {
    public int[] solution(int[] fees, String[] records) {
        
        Fee fee = new Fee(fees[0], fees[1], fees[2], fees[3]);

        TreeMap<String, Car> cars = new TreeMap<>();

        for (String record : records) {
            String[] tokens = record.split(" ");

            int time = parseTime(tokens[0]);
            String carNumber = tokens[1];
            boolean isIn = tokens[2].equals("IN");

            cars.putIfAbsent(carNumber, new Car());

            Car car = cars.get(carNumber);

            if (isIn) {
                car.in(time);
            } else {
                car.out(time);
            }
        }

        int endTime = parseTime("23:59");
        for (Car car : cars.values()) {
            car.out(endTime);
        }

        int[] ans = new int[cars.size()];

        for (int i = 0; i < ans.length; i++) {

            Map.Entry<String, Car> entry = cars.pollFirstEntry();
            Car car = entry.getValue();

            ans[i] = car.getCost(fee);
        }

        return ans;
    }
    
    public int parseTime(String s) {
        int hour = Integer.parseInt(s.substring(0, 2));
        int minute = Integer.parseInt(s.substring(3));

        return hour * 60 + minute;
    }
    
    static class Fee {

        int baseTime, baseFee;
        int unitTime, unitFee;

        public Fee(int baseTime, int baseFee, int unitTime, int unitFee) {
            this.baseTime = baseTime;
            this.baseFee = baseFee;
            this.unitTime = unitTime;
            this.unitFee = unitFee;
        }

        public int calculateCost(int time) {

            int fee = baseFee;

            int exceedTime = Math.max(0, time - baseTime);
            fee += Math.ceil(exceedTime * 1.0 / unitTime) * unitFee;

            return fee;
        }
    }

    static class Car {

        int lastInTime;
        int totalTime;

        public void in(int inTime) {
            this.lastInTime = inTime;
        }

        public void out(int outTime) {
            if (lastInTime == -1) return;

            this.totalTime += (outTime - lastInTime);
            this.lastInTime = -1;
        }

        public int getCost(Fee fee) {
            return fee.calculateCost(totalTime);
        }
    }
}